### PR TITLE
Added paging functionality to support reading very large, or wide, data files

### PR DESCRIPTION
**Pull request by ChrisMarsh82**
Added Model and Databook folder and created model classes for the databook and dataframes.

### CODE REVIEW

**Comment by lloyddewit** [comment]
> 
> 
> Added Model and Databook folder and created model classes for the databook and dataframes.

@ChrisMarsh82 Thanks for the PR. I liked how you labelled your commits. Tips for future PRs:
- When you open a PR its always good to give as much information as possible about the issue you're trying to solve.
- If this PR resolves an existing issue (e.g. issue #4832), then you should write 'Fixes issue #4832. GitHub will then automatically close the issue when this PR is merged.
- You shouldn't mention the issue number in the PR title
- The PR title is automatically included in the release notes provided to users. Therefore the title should explain (from a user perspective) what has changed. I updated the title but please feel free to improve

**Comment by lloyddewit** [comment]
@ChrisMarsh82 I don't think the files below should be included in this PR. Can we remove them (I can help if needed)?
```
 instat/packages.config
 instat/translations/en/r_instat_not_menus.json
 instat/translations/rInstatTranslations.db
 instat/translations/translationsManuallyEntered.csv
 ...ges/TranslateWinForm.1.1.3/.signature.p7s → ...ges/TranslateWinForm.1.1.2
 packages/TranslateWinForm.1.1.2/lib/net461/TranslateWinForm.dll
 ...orm.1.1.3/lib/net461/TranslateWinForm.xml → ...orm.1.1.2/lib/net461/TranslateWinForm.xml
 packages/TranslateWinForm.1.1.3/lib/net461/TranslateWinForm.dll
```

**Comment by lloyddewit** [comment]
@ChrisMarsh82 I get some build warnings
image

**Comment by lloyddewit** [comment]
@ChrisMarsh82 I see that you created a folder structure for the VB files. I can see the advantages of structuring the files in this way. At some point, we should probably do this for all the source files in R-Instat.
The other VB files are currently in a flat structure. If we use a hierarchy, then I think we should do this consistently across the code base. Until then, I suggest we use the flat structure, else it gets confusing. We can discuss with @dannyparsons .

File: instat/clsGridLink.vb [hunk 1]
```diff
@@ -48,39 +48,48 @@ Public Class clsGridLink
...
-        clsGetDataNames.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$get_data_names")
-        clsGetDataNames.AddParameter("include_hidden", "FALSE")
-        clsGetDataFrame.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$get_data_frame")
+    Private Sub UpdateGrids()
+        'TODO remove this sub and whole class if possible
+
+        frmMain.ucrDataViewer.RefreshGridData()
+        frmMain.ucrColumnMeta.RefreshGridData()
+        frmMain.ucrDataFrameMeta.RefreshGridData()
+        Exit Sub
```
Comment (lloyddewit) [inline instat/clsGridLink.vb:57]: I assume that you'll delete the rest of this sub, is that correct?

File: instat/clsGridLink.vb [hunk 2]
```diff
@@ -212,7 +221,7 @@ Public Class clsGridLink
                                 Else
                                     FillSheet(dfTemp, strDataName, grdData, bInstatObjectDataFrame:=True, bIncludeDataTypes:=True, iNewPosition:=i, bFilterApplied:=False, bCheckFreezeColumns:=True)
                                 End If
-                                ucrDataViewer.SetColumnNames(strDataName, dfTemp.ColumnNames())
+                                '  ucrDataViewer.SetColumnNames(strDataName, dfTemp.ColumnNames())
```
Comment (lloyddewit) [inline instat/clsGridLink.vb:224]: I assume that you'll delete this line?

File: instat/clsGridLink.vb [hunk 3]
```diff
@@ -271,21 +280,21 @@ Public Class clsGridLink
...
-            grdMetadata.Visible = False
-        Else
-            grdData.Visible = True
-            grdVariablesMetadata.Visible = True
-            grdMetadata.Visible = True
-        End If
-        ucrDataViewer.UpdateCurrentWorksheet()
-    End Sub
-
+        If grdData.Worksheets.Count = 0 Then
```
Comment (lloyddewit) [inline instat/clsGridLink.vb:286]: This and all subsequent changes in this file are not real changes, the GitHub diff seems to be confused (or non-visible characters were inserted accidentally)

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@ChrisMarsh82 I didn't really look at the contents of individual files in this review, I'll do that later. 
I added some higher level comments to the PR. There are also some merge conflicts.
The easiest way to resolve these comments and conflcts could be to make a new clean PR. We can discuss if needed.

**Comment by ChrisMarsh82** [comment]
@lloyddewit I will try to sort conflicts and let you know if I need your help. May need to have another look at some of the code as my original plan was to remove clsLink completely. I seem to remember that was a grid that still used it but it may be worth me just changing that and removing the file altogether.

**Comment by ChrisMarsh82** [comment]
@lloyddewit @dannyparsons Regarding the file structure change. I feel we need to start doing this but it cannot be done in one pass. There is a problem with circular references in most of the files. My plan was to slowly move files into a file structure that don't have 'bad' references. That way we can add namespaces once all files are in structured folders.

**Comment by ChrisMarsh82** [comment]
@lloyddewit Sorry about the confusion on this one. I have now done it again on a completely new pull request see https://github.com/africanmathsinitiative/R-Instat/pull/6707
The new one is much cleaner.

**Pull request closed without merging**